### PR TITLE
[DOC-2550] - Code coverage and artifact tab plugin

### PR DESCRIPTION
**Pull request by aimurphy**
## Code Coverage

- New topic about code coverage in CI pipelines.
- Preview - Code Coverage: https://64837a5be88c67085e2cd93b--harness-developer.netlify.app/docs/continuous-integration/use-ci/set-up-test-intelligence/code-coverage

## Artifacts tab

- New section in multiple topics for using the artifacts-metadata-publisher plugin.
- Reorg **Upload Artifacts to S3** and **Upload Artifacts to GCS** so they match the structure of **Upload Artifacts to JFrog Artifactory**.
- Preview - S3: https://64837a5be88c67085e2cd93b--harness-developer.netlify.app/docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings
- Preview - GCS: https://64837a5be88c67085e2cd93b--harness-developer.netlify.app/docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-gcs-step-settings
- Preview - JFrog: https://64837a5be88c67085e2cd93b--harness-developer.netlify.app/docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-jfrog
- Preview - View tests: https://64837a5be88c67085e2cd93b--harness-developer.netlify.app/docs/continuous-integration/use-ci/set-up-test-intelligence/viewing-tests#publish-reports-to-the-artifacts-tab

## Allure tutorial

Renamed and moved to appropriate category: https://64837a5be88c67085e2cd93b--harness-developer.netlify.app/tutorials/ci-pipelines/publish/artifacts-tab

### CODE REVIEW

File: docs/continuous-integration/use-ci/set-up-test-intelligence/code-coverage.md [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+   * [Upload Artifacts to S3](../build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md)
+
+3. Optional: Add a step to [publish your code coverage report to the Artifacts tab](#publish-code-coverage-reports-to-the-artifacts-tab).
+
+### Java
+
+1. Set up a Java code coverage tool, such as [JaCoCo](https://github.com/jacoco/jacoco).
+2. Run your tests in a **Run** or **Run Tests** step.
+
+   By including JaCoCo in `pom.xml`, the `mnv test` command automatically writes a code coverage report to an `exec` file.
```
Comment (jimsheldon) [inline docs/continuous-integration/use-ci/set-up-test-intelligence/code-coverage.md:46]: I think this should be `mvn test` right?

Comment (aimurphy) [reply docs/continuous-integration/use-ci/set-up-test-intelligence/code-coverage.md:46]: yes, thank you, that's a typo. This is why we have reviews lol :)

**Comment by rustd** [comment]
@aimurphy thank you for writing this. Some feedback below.
- For all language examples, please add the yaml example, similar to JavaScript one.
- Add a step to upload your code coverage report to cloud storage - Remove "Upload Artifacts to JFrog Artifactory" from all the languages
-  Remove this from the yaml files connectorRef: account.harnessImage
- Remove the Visual tab and just keep it as yaml only.
- Remove optional verbiage from Optional: Add a step to [publish your code coverage report to the Artifacts tab]
- For uploadtos3, please refer to this plugin https://github.com/harness-community/drone-s3-upload-publish
- For Jacoco, please add this plugin link https://github.com/harness-community/drone-jacoco-s3
- Rename "Publish code coverage reports to the Artifacts tab" to "View code coverage reports in the Artifacts tab"
- Rename "Publish an Allure Report to the Artifacts tab" to "Publish any URL link in the Artifacts tab"

**Comment by aimurphy** [comment]
@rustd 

_~For all language examples, please add the yaml example, similar to JavaScript one.~_

_~Add a step to upload your code coverage report to cloud storage - Remove "Upload Artifacts to JFrog Artifactory" from all the languages~_

_Remove this from the yaml files connectorRef: account.harnessImage._ **Partially done:** `connectorRef` is required for **Plugin** step even on Hosted infra. I removed it from the python yaml example, but I need to keep it in the artifact-metadata-publisher plugin example.

_Remove the Visual tab and just keep it as yaml only._ **Won't do:** Since we provide two types of editors, we need to provide instructions for both. Unless there is some configuration here that is YAML-only (ex. as with Cache Intelligence), we need to keep the visual editor steps. The YAML tab is set to default, so it is automatically selected when users browse to this page. Technically I should provide visual & yaml steps for all the languages too, but those instructions are higher level than the plugin config steps.

_~Remove optional verbiage from Optional: Add a step to [publish your code coverage report to the Artifacts tab]~_

_~For uploadtos3, please refer to this plugin https://github.com/harness-community/drone-s3-upload-publish~_

_For Jacoco, please add this plugin link https://github.com/harness-community/drone-jacoco-s3_ **Done with comment:** Previously, I had the JaCoCo plugin in a Tip box. We can't assume they are using JaCoCo, but I reorganized the Java instructions to describe both options equally.

_~Rename "Publish code coverage reports to the Artifacts tab" to "View code coverage reports in the Artifacts tab"~_

_~Rename "Publish an Allure Report to the Artifacts tab" to "Publish any URL link in the Artifacts tab"~_

**Comment by rustd** [comment]
Thanks! Looks great
- Remove JaCoCoplugin section from "View code coverage reports on the Artifacts tab"

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-gcs-step-settings.md [hunk 2]
```diff
@@ -57,3 +67,60 @@ Set the timeout limit for the step. Once the timeout limit is reached, the step
 
 * [Step Skip Condition settings](../../../platform/8_Pipelines/w_pipeline-steps-reference/step-skip-condition-settings.md)
 * [Step Failure Strategy settings](../../../platform/8_Pipelines/w_pipeline-steps-reference/step-failure-strategy-settings.md)
+
+## Confirm the upload
+
+After adding the steps and saving the pipeline, select **Run** to run the pipeline.
```
Comment (brian-f-harness) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-gcs-step-settings.md:73]:
```suggestion
After you add the steps and save the pipeline, select **Run** to run the pipeline.
```

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md [hunk 3]
```diff
@@ -99,3 +122,134 @@ Set the timeout limit for the step. Once the timeout limit is reached, the step
 
 * [Step Skip Condition settings](../../../platform/8_Pipelines/w_pipeline-steps-reference/step-skip-condition-settings.md)
 * [Step Failure Strategy settings](../../../platform/8_Pipelines/w_pipeline-steps-reference/step-failure-strategy-settings.md)
+
+## Confirm the upload
+
+After adding the steps and saving the pipeline, select **Run** to run the pipeline.
```
Comment (brian-f-harness) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md:128]:
```suggestion
After you add the steps and save the pipeline, select **Run** to run the pipeline.
```

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md [hunk 4]
```diff
@@ -99,3 +122,134 @@ Set the timeout limit for the step. Once the timeout limit is reached, the step
...
+
+Add a [Plugin step](../use-drone-plugins/plugin-step-settings-reference.md) that uses the `drone-s3-upload-publish` plugin.
+
+Configure the **Plugin** step settings as follows:
+
+   * **Name:** Enter a name.
+   * **Container Registry:** Select a Docker connector.
+   * **Image:** Enter `harnesscommunity/drone-s3-upload-publish`.
+   * **Settings:** Add the following seven settings as key-value pairs.
+      * `aws_access_key_id`: An [expression](/docs/platform/references/runtime-inputs/#expressions) referencing a [Harness secret](/docs/category/secrets) or [pipeline variable](/docs/platform/Variables-and-Expressions/add-a-variable) containing your AWS access ID, such as `<+pipeline.variables.AWS_ACCESS>`
```
Comment (brian-f-harness) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md:208]:
```suggestion
      * `aws_access_key_id`: An [expression](/docs/platform/references/runtime-inputs/#expressions) referencing a [Harness secret](/docs/category/secrets) or [pipeline variable](/docs/platform/Variables-and-Expressions/add-a-variable) containing your AWS access ID, such as `<+pipeline.variables.AWS_ACCESS>`.
```

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md [hunk 5]
```diff
@@ -99,3 +122,134 @@ Set the timeout limit for the step. Once the timeout limit is reached, the step
...
+      * `aws_secret_access_key`: An [expression](/docs/platform/references/runtime-inputs/#expressions) referencing a [Harness secret](/docs/category/secrets) or [pipeline variable](/docs/platform/Variables-and-Expressions/add-a-variable) containing your AWS access key, such as `<+pipeline.variables.AWS_SECRET>`
```
Comment (brian-f-harness) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md:209]:
```suggestion
      * `aws_secret_access_key`: An [expression](/docs/platform/references/runtime-inputs/#expressions) referencing a [Harness secret](/docs/category/secrets) or [pipeline variable](/docs/platform/Variables-and-Expressions/add-a-variable) containing your AWS access key, such as `<+pipeline.variables.AWS_SECRET>`.
```

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md [hunk 6]
```diff
@@ -99,3 +122,134 @@ Set the timeout limit for the step. Once the timeout limit is reached, the step
...
+      * `aws_default_region`: Your default AWS region, such as `ap-southeast-2`
```
Comment (brian-f-harness) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md:210]:
```suggestion
      * `aws_default_region`: Your default AWS region, such as `ap-southeast-2`.
```

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md [hunk 7]
```diff
@@ -99,3 +122,134 @@ Set the timeout limit for the step. Once the timeout limit is reached, the step
...
+      * `aws_bucket`: The target S3 bucket.
+      * `artifact_file`: `url.txt`
```
Comment (brian-f-harness) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md:212]:
```suggestion
      * `artifact_file`: `url.txt`.
```

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md [hunk 8]
```diff
@@ -99,3 +122,134 @@ Set the timeout limit for the step. Once the timeout limit is reached, the step
...
+      * `source`: The path to store and retrieve the artifact in the S3 bucket.
+   * **Image Pull Policy:** Select **If Not Present**
```
Comment (brian-f-harness) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md:214]:
```suggestion
   * **Image Pull Policy:** Select **If Not Present**.
```

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md [hunk 9]
```diff
@@ -99,3 +122,134 @@ Set the timeout limit for the step. Once the timeout limit is reached, the step
...
+```
+
+Configure the **Plugin** step settings as follows:
+
+* **Name:** Enter a name.
+* **Container Registry:** Select a Docker connector.
+* **Image:** Enter `plugins/artifact-metadata-publisher`.
+* **Settings:** Add the following two settings as key-value pairs.
+  * `file_urls`: The URL to the target artifact that was uploaded in the **Upload Artifacts to S3** step.
+  * `artifact_file`: `artifact.txt`
```
Comment (brian-f-harness) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md:159]:
```suggestion
  * `artifact_file`: `artifact.txt`.
```

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-jfrog.md [hunk 10]
```diff
@@ -67,24 +66,65 @@ Set the timeout limit for the step. Once the timeout limit is reached, the step
...
+```
+
+Configure the **Plugin** step settings as follows:
+
+* **Name:** Enter a name.
+* **Container Registry:** Select a Docker connector.
+* **Image:** Enter `plugins/artifact-metadata-publisher`.
+* **Settings:** Add the following two settings as key-value pairs.
+  * `file_urls`: The URL to the target artifact that was uploaded in the **Upload Artifacts to JFrog Artifactory** step.
+  * `artifact_file`: `artifact.txt`
```
Comment (brian-f-harness) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-jfrog.md:101]:
```suggestion
  * `artifact_file`: `artifact.txt`.
```

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-gcs-step-settings.md [hunk 11]
```diff
@@ -57,3 +67,60 @@ Set the timeout limit for the step. Once the timeout limit is reached, the step
...
+  <TabItem value="Visual" label="Visual">
+```
+
+Configure the **Plugin** step settings as follows:
+
+* **Name:** Enter a name.
+* **Container Registry:** Select a Docker connector.
+* **Image:** Enter `plugins/artifact-metadata-publisher`.
+* **Settings:** Add the following two settings as key-value pairs.
+  * `file_urls`: A GCS URL using the **Bucket** and **Target** specified in the **Upload Artifacts to GCS** step, such as `https://storage.googleapis.com/GCS_BUCKET_NAME/TARGET_PATH.html`
```
Comment (brian-f-harness) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-gcs-step-settings.md:98]:
```suggestion
  * `file_urls`: A GCS URL using the **Bucket** and **Target** specified in the **Upload Artifacts to GCS** step, such as `https://storage.googleapis.com/GCS_BUCKET_NAME/TARGET_PATH.html`.
```

File: docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-gcs-step-settings.md [hunk 12]
```diff
@@ -57,3 +67,60 @@ Set the timeout limit for the step. Once the timeout limit is reached, the step
...
+  * `artifact_file`: `artifact.txt`
```
Comment (brian-f-harness) [inline docs/continuous-integration/use-ci/build-and-upload-artifacts/upload-artifacts-to-gcs-step-settings.md:99]:
```suggestion
  * `artifact_file`: `artifact.txt`.
```

File: docs/continuous-integration/use-ci/set-up-test-intelligence/viewing-tests.md [hunk 13]
```diff
@@ -12,14 +12,67 @@ Your pipelines can run tests in **Run** and **Run Tests** steps.
...
+1. Include steps in your pipeline that run tests and produce test reports.
+2. Include a step to upload the report artifact to cloud storage, such as [Upload Artifacts to GCS](../build-and-upload-artifacts/upload-artifacts-to-gcs-step-settings.md), [Upload Artifacts to S3](../build-and-upload-artifacts/upload-artifacts-to-s-3-step-settings.md), or [Upload Artifacts to JFrog Artifactory](../build-and-upload-artifacts/upload-artifacts-to-jfrog.md).
+3. Include a [Plugin step](../use-drone-plugins/plugin-step-settings-reference.md) that uses the `artifact-metadata-publisher` plugin. Configure the **Plugin** step settings as follows:
+
+   * **Name:** Enter a name.
+   * **Container Registry:** Select a Docker connector.
+   * **Image:** Enter `plugins/artifact-metadata-publisher`.
+   * **Settings:** Add the following two settings as key-value pairs.
+      * `file_urls`: The URL to the target artifact that was uploaded in the **Upload Artifacts** step.
+      * `artifact_file`: `artifact.txt`
```
Comment (brian-f-harness) [inline docs/continuous-integration/use-ci/set-up-test-intelligence/viewing-tests.md:51]:
```suggestion
      * `artifact_file`: `artifact.txt`.
```

**Review by brian-f-harness — APPROVED** [review]
lgtm. Added minor suggestions.

**Pull request merged**